src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs

```diff
@@ -192,52 +193,55 @@ public struct AsyncTaskMethodBuilder<TResult>
-            Debugger.NotifyOfCrossThreadDependency();
-
-            // At this point, taskField should really be null, in which case we want to create the box.
-            // However, in a variety of debugger-related (erroneous) situations, it might be non-null,
-            // e.g. if the Task property is examined in a Watch window, forcing it to be lazily-initialized
-            // as a Task<TResult> rather than as an AsyncStateMachineBox.  The worst that happens in such
-            // cases is we lose the ability to properly step in the debugger, as the debugger uses that
-            // object's identity to track this specific builder/state machine.  As such, we proceed to
-            // overwrite whatever's there anyway, even if it's non-null.
+            else
+            {
+                // Alert a listening debugger that we can't make forward progress unless it slips threads.
+                // If we don't do this, and a method that uses "await foo;" is invoked through funceval,
+                // we could end up hooking up a callback to push forward the async method's state machine,
+                // the debugger would then abort the funceval after it takes too long, and then continuing
+                // execution could result in another callback being hooked up.  At that point we have
+                // multiple callbacks registered to push the state machine, which could result in bad behavior.
+                Debugger.NotifyOfCrossThreadDependency();
+
+                // At this point, taskField should really be null, in which case we want to create the box.
+                // However, in a variety of debugger-related (erroneous) situations, it might be non-null,
+                // e.g. if the Task property is examined in a Watch window, forcing it to be lazily-initialized
+                // as a Task<TResult> rather than as an AsyncStateMachineBox.  The worst that happens in such
+                // cases is we lose the ability to properly step in the debugger, as the debugger uses that
+                // object's identity to track this specific builder/state machine.  As such, we proceed to
+                // overwrite whatever's there anyway, even if it's non-null.
 #if NATIVEAOT
-            // DebugFinalizableAsyncStateMachineBox looks like a small type, but it actually is not because
-            // it will have a copy of all the slots from its parent. It will add another hundred(s) bytes
-            // per each async method in NativeAOT binaries without adding much value. Avoid
-            // generating this extra code until a better solution is implemented.
-            var box = new AsyncStateMachineBox<TStateMachine>();
+                // DebugFinalizableAsyncStateMachineBox looks like a small type, but it actually is not because
+                // it will have a copy of all the slots from its parent. It will add another hundred(s) bytes
```

Comment:
It might be out of scope for this PR, but how about using an unspecialized `DebugFinalizableAsyncStateMachineBox<IAsyncStateMachine>` for NativeAOT?

Reply:
Don't know much about this - if it's only used by managed debuggers, it would not be useful for native AOT anyway. It's ifdeffed out so I'm not concerned about it.